### PR TITLE
Disallow Terraform from touching state from future TF versions

### PR DESCRIPTION
**Pull request by mitchellh**
Not sure if this closes any existing issues!

This adds a field `terraform_version` to the state that represents the Terraform version that wrote that state. If Terraform encounters a state written by a future version, it will error. You must use at least the version that wrote that state.

Internally we have fields to override this behavior (`StateFutureAllowed`), but I chose not to expose them as CLI flags, since the user can just modify the state directly. This is tricky, but should be tricky to represent the horrible disaster that can happen by enabling it.

We didn't have to bump the state format version since the absense of the field means it was written by version "0.0.0" which will always be older.

### CODE REVIEW

**Comment by pearkes** [comment]
If the version is say:

> Terraform v0.6.12-dev (938ab99d51195f04fbde54d7bcc9f2e53e0ffcfc+CHANGES)

What will the behavior be? Will it pin against 0.6.12 or the specific sha?

**Comment by mitchellh** [comment]
@pearkes It will be 0.6.12. We follow semver directly for comparisons. The SHA means nothing, the `-dev` just means "an early version of 0.6.12": 0.6.12 > 0.6.12-dev > 0.6.11

**Comment by mitchellh** [comment]
The test failure there is now seems unrelated. I'm not sure why that is failling.

**Comment by mitchellh** [comment]
Finally fixed the tests. They actually caught a very significant issue around the way Atlas would check this for state conflicts. I've since changed the impl around significantly to get these tests to pass (without touching them).

File: terraform/context.go [hunk 1]
```diff
@@ -92,6 +93,24 @@ func NewContext(opts *ContextOpts) *Context {
...
+	// If our state is from the future, then error. Callers can avoid
+	// this error by explicitly setting `StateFutureAllowed`.
+	if !opts.StateFutureAllowed && state.FromFutureTerraform() {
+		return nil, fmt.Errorf(
+			"Terraform doesn't allow running any operations against a state\n"+
+				"that was written by a future Terraform version. The state is\n"+
+				"reporting it is written by Terraform '%s'.\n\n"+
+				"Please run at least that version of Terraform to continue. Many\n"+
+				"command line operations also provide a flag to override this\n"+
+				"protection",
```
Comment (phinze) [inline terraform/context.go]: Based on the PR description, this last sentence is not accurate. Sounds like we're going for manual edit-the-state override in this case?

If this is the case, I wonder if the thing to do here is add a docs page and link off to it in the error message?

Comment (mitchellh) [inline terraform/context.go]: This is correct. I'll modify this message.

Comment (phinze) [inline terraform/context.go]: Ping :grin: :sparkles:

**Comment by mitchellh** [comment]
Fixed the feedback. Going to rebase this onto dev-0.7

**Comment by mitchellh** [comment]
Going to reopen PR against 0.7

**Pull request closed without merging**